### PR TITLE
Add directives and pipes for CourseItem component.

### PR DESCRIPTION
**Pull request by VitalySokolov**

### CODE REVIEW

File: src/app/video-courses/course-item/course-item.component.host.spec.ts [hunk 1]
```diff
@@ -34,13 +36,13 @@ class TestHostComponent {
...
 
-describe('CourseItemComponent with Test Host', () => {
+describe('CourseItemComponent with TestHost', () => {
   let testHost: TestHostComponent;
   let fixture: ComponentFixture<TestHostComponent>;
 
   beforeEach(() => {
     TestBed.configureTestingModule({
-      declarations: [CourseItemComponent, TestHostComponent],
+      declarations: [CourseItemComponent, TestHostComponent, CourseDurationPipe, HighlightCourseDirective],
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-item/course-item.component.host.spec.ts]: Update formatting

File: src/app/video-courses/course-item/course-item.component.spec.ts [hunk 2]
```diff
@@ -51,12 +53,18 @@ describe('CourseItemComponent', () => {
...
   it('should display course information', () => {
-    expect(title.nativeElement.innerText).toBe(courseTitle);
+    expect(title.nativeElement.innerText).toBe(courseTitle.toUpperCase());
     expect(description.nativeElement.innerText).toBe(courseDescription);
   });
 
   it('should emit correct course id when click "Delete" button', () => {
     let deletedCourseId: number;
+    let date = new Date();
+    console.log(date.valueOf());
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-item/course-item.component.spec.ts]: Remove logs

File: src/app/video-courses/course-item/course-item.component.ts [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
   @Input() courseTitle: string;
   @Input() courseDescription: string;
-  @Input() courseDate: Date;
+  @Input() courseDateMillis: number;
+  courseDate: Date;
   @Input() courseDuration: number;
   @Output() courseDeleted = new EventEmitter<number>();
+
+  ngOnInit(): void {
+    this.courseDate = new Date(this.courseDateMillis);
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-item/course-item.component.ts]: Better to transform timestamp to `Date` object inside service. If you want to keep it in component better to transform in property setter instead.

File: src/app/video-courses/course-list/course-list.component.html [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
   [courseTitle]="course.title"
   [courseDescription]="course.description"
-  [courseDate]="course.date"
+  [courseDateMillis]="course.date"
   [courseDuration]="course.duration"
   (courseDeleted)="onCourseDeleted(course.id)">
 </app-course-item>
-<button class="btn btn-info btn-block mt-3" (click)="loadMoreCourses()">Load More</button>
+<button class="btn btn-info btn-block mt-3" (click)="loadMoreCourses()">
+  <span *ngIf="courseList.length">Load More</span>
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-list/course-list.component.html]: use `*ngIf` + `else` syntax instead 2 `ngIf`

File: src/app/video-courses/course-list/course-list.component.spec.ts [hunk 5]
```diff
@@ -14,21 +15,21 @@ describe('CourseListComponent', () => {
       id: 123,
       title: 'Test Course 1',
       description: 'Description 1',
-      date: new Date(),
+      date: (new Date()).valueOf() - 500000000,
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-list/course-list.component.spec.ts]: Use `(new Date()).getTime()` as more convinient way to get timestamp in JS. `valueOf` is polymorphic method existing on a variety of Objects in JS, that used in comparison algorithm.

File: src/app/video-courses/filter-by-title.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { VideoCourseItem } from './video-course-item.model';
+
+@Pipe({
+  name: 'filterByTitle'
+})
+export class FilterByTitlePipe implements PipeTransform {
+
+  transform(allCourses: VideoCourseItem[], searchString: string): VideoCourseItem[] {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/filter-by-title.pipe.ts]: Make less specific, you can filter any entity by title field, not only `VideoCourseItem`. For an input value type use something like `Partial<{title: string}>` as inside pipe you care only about `title` field.

File: src/app/video-courses/highlight-course.directive.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+@Directive({
+  selector: '[appHighlightCourse]'
+})
+export class HighlightCourseDirective implements OnInit {
+  @Input('appHighlightCourse') courseDate: Date;
+
+  private static isFreshCourse(courseDate: Date, currentDate: Date) {
+    const millisecondsInDay = 1000 * 60 * 60 * 24;
+    const twoWeeks = 14;
+    const diffInDays = (currentDate.valueOf() - courseDate.valueOf()) / millisecondsInDay;
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/highlight-course.directive.ts]: `getTime` as written before.

File: src/app/video-courses/highlight-course.directive.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { Directive, ElementRef, Input, OnInit } from '@angular/core';
+
+@Directive({
+  selector: '[appHighlightCourse]'
+})
+export class HighlightCourseDirective implements OnInit {
+  @Input('appHighlightCourse') courseDate: Date;
+
+  private static isFreshCourse(courseDate: Date, currentDate: Date) {
+    const millisecondsInDay = 1000 * 60 * 60 * 24;
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/highlight-course.directive.ts]: Move to static prop, not calculate on each static method invocation.

File: src/app/video-courses/highlight-course.directive.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    return diffInDays <= twoWeeks;
+  }
+
+  constructor(private el: ElementRef) {
+  }
+
+  ngOnInit() {
+    const currentDate = new Date();
+
+    if (currentDate < this.courseDate) {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/highlight-course.directive.ts]: wrong comparison, use `currentDate.getTime() < this.courseDate.getTime()`

File: src/app/video-courses/highlight-course.directive.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      this.highlightFutureCourse();
+    } else if (HighlightCourseDirective.isFreshCourse(this.courseDate, currentDate)) {
+      this.highlightFreshCourse();
+    }
+  }
+
+  private highlightFutureCourse() {
+    this.el.nativeElement.style.border = `3px solid blue`;
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/highlight-course.directive.ts]: IMHO, better to add transparent border `OnInit` and then just set the border color, in this case, you'll not trigger layout in CSS

File: src/app/video-courses/search-course/search-course.component.ts [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
   selector: 'app-search-course',
   templateUrl: './search-course.component.html',
   styleUrls: ['./search-course.component.css']
 })
 export class SearchCourseComponent implements OnInit {
+  @Output() searchCourse = new EventEmitter<string>();
+
   search = '';
 
   constructor() {
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/search-course/search-course.component.ts]: Remove empty constructor

File: src/app/video-courses/course-item/course-item.component.host.spec.ts (hunk 1 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-item/course-item.component.host.spec.ts]: Done

File: src/app/video-courses/course-item/course-item.component.spec.ts (hunk 2 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-item/course-item.component.spec.ts]: Done

File: src/app/video-courses/course-list/course-list.component.html (hunk 4 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-list/course-list.component.html]: Done

File: src/app/video-courses/search-course/search-course.component.ts (hunk 11 above)
Comment (VitalySokolov) [reply src/app/video-courses/search-course/search-course.component.ts]: Done

File: src/app/video-courses/highlight-course.directive.ts (hunk 8 above)
Comment (VitalySokolov) [reply src/app/video-courses/highlight-course.directive.ts]: Done

File: src/app/video-courses/filter-by-title.pipe.ts (hunk 6 above)
Comment (VitalySokolov) [reply src/app/video-courses/filter-by-title.pipe.ts]: Done

File: src/app/video-courses/course-item/course-item.component.ts (hunk 3 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-item/course-item.component.ts]: Moved to service.

File: src/app/video-courses/course-list/course-list.component.spec.ts (hunk 5 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-list/course-list.component.spec.ts]: Done

File: src/app/video-courses/highlight-course.directive.ts (hunk 7 above)
Comment (VitalySokolov) [reply src/app/video-courses/highlight-course.directive.ts]: Done

File: src/app/video-courses/highlight-course.directive.ts (hunk 9 above)
Comment (VitalySokolov) [reply src/app/video-courses/highlight-course.directive.ts]: Done

File: src/app/video-courses/course-item/course-item.component.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
-import { Component, EventEmitter, Input, Output } from '@angular/core';
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/course-item/course-item.component.ts]: Remove unused `OnInit`

File: src/app/video-courses/highlight-course.directive.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    const diffInDays = (currentDate.getTime() - courseDate.getTime()) / HighlightCourseDirective.millisecondsInDay;
+
+    return diffInDays <= twoWeeks;
+  }
+
+  constructor(private el: ElementRef) {
+  }
+
+  ngOnInit() {
+    this.el.nativeElement.style.border = `3px solid transparent`;
```
Comment (vitaliy-bobrov) [inline src/app/video-courses/highlight-course.directive.ts]: no need to use ```

File: src/app/video-courses/course-item/course-item.component.ts (hunk 12 above)
Comment (VitalySokolov) [reply src/app/video-courses/course-item/course-item.component.ts]: Done.

File: src/app/video-courses/highlight-course.directive.ts (hunk 13 above)
Comment (VitalySokolov) [reply src/app/video-courses/highlight-course.directive.ts]: Done.

**Pull request merged**